Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -397,33 +485,50 @@ extension SwiftLanguageServer {
     let keys = self.keys
 
     self.queue.async {
+      let uri = note.textDocument.uri
+      let newVersion = note.textDocument.version ?? -1
       var lastResponse: SKDResponseDictionary? = nil
 
-      let snapshot = self.documentManager.edit(note) { (before: DocumentSnapshot, edit: TextDocumentContentChangeEvent) in
-        let req = SKDRequestDictionary(sourcekitd: self.sourcekitd)
-        req[keys.request] = self.requests.editor_replacetext
-        req[keys.name] = note.textDocument.uri.pseudoPath
+      // We iterate manually over the `contentChanges` here instead of passing
+      // the notification directly to the convenience method `edit(_:editCallback:)`
+      // since we need to update the tokens in lockstep with the edits in the
+      // document.
```

Comment:
Currently we expect all changes to lexical and syntactic tokens to go through `updateLexicalAndSyntacticTokens` on `SwiftLanguageServer`'s queue, which synchronously funnels them through `DocumentManager`'s `replaceLexical/SyntacticTokens` on its own queue.

If we were to update the tokens through modified snapshots, I am worried that we may end up duplicating a lot of logic that previously was cleanly categorized into the `replaceXTokens` functions in `DocumentManager`.

If we were to stay on `DocumentManager`'s queue by calling `updateLexicalAndSyntacticTokens` from the edit callback, we would have to weaken the restriction that `updateLexicalAndSyntacticTokens` runs on `SwiftLanguageServer`'s queue and be careful not to run into deadlocks with `DocumentManager`'s queue when calling `replaceLexical/SyntacticTokens`. Note that `openDocument` would still call `updateLexicalAndSyntacticTokens` in that case.

Not sure what a good solution would be here, either one feels a bit unsatisfying.

Reply:
What do you think about the following: 

The only instance member that `updateLexicalAndSyntacticTokens` accesses is `documentManager`, so it’s not really accessing any state private to `SwiftLanguageServer` and doesn’t really need to run on the server’s queue.
If we now change `updateLexicalAndSyntacticTokens` to return `DocumentTokens` instead of updating the `documentManager`, we can make `updateLexicalAndSyntacticTokens` static and definitely remove the requirement to run on `SwiftLanguageServer`’s queue.

Now, the `editCallback` could return `DocumentTokens` to the document manager, which could incorporate them into the current document.

So, we’d have method signatures like:
```swift
// SwiftLanguageServer
private static func updateLexicalAndSyntacticTokens(response: SKDResponseDictionary, for snapshot: DocumentSnapshot) -> DocumentTokens

// DocumentManager
public func edit(_ uri: DocumentURI, newVersion: Int, edits: [TextDocumentContentChangeEvent], editCallback: ((_ before: DocumentSnapshot, TextDocumentContentChangeEvent) -> DocumentTokens?)? = nil) throws -> DocumentSnapshot
```

Reply:
Should we do this for the other update methods too, to be consistent?

Reply:
I think it has to be an instance method, since we need `sourcekitd.keys`.